src/halmos/solvers.py

```diff
@@ -475,6 +475,7 @@ def find_solver_binary(solver: SolverInfo) -> Path | None:
     return None
 
 
+@synchronized()
```

Comment:
i guess the performance impact of introducing a lock here shouldn't be a concern, because  find_solver_binary should be quick, so the lock wait time for other processes would be negligible, right?

that said, i'm wondering if we could move the synchronized decorator to install_solver() to minimize any potential performance overhead.

while this function is always executed, install_solver is typically not. so if we add a check at the beginning of install_solver (possibly by calling find_solver_binary again) to see whether the solver is already installed, we could safely apply the synchronized decorator there. this way, locking would be avoided in most cases.